bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/update/update.xml

```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
+<update:update-descriptions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:update="https://openhab.org/schemas/update-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/update-description/v1.0.0 https://openhab.org/schemas/update-description-1.0.0.xsd">
+
+	<thing-type uid="mystrom:mystromplug">
+		<instruction-set targetVersion="1">
```

Comment:
The `thingTypeVersion` property needs to be set accordingly.

Reply:
Please see my latest commit. Is the version property set correct? I looked at other ".../update/instructions.xml" files but couldnt find the version property set anywhere in them

Reply:
You need to set it for the corresponding thing, so probably here:
https://github.com/openhab/openhab-addons/blob/e1c9213cfcdc094f290c8ff7a6e5a2b5b1796c9e/bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml#L17-L28

See also for example #14572.

Reply:
Thanks for this example, I have changed the files accordingly.

Reply:
Can you confirm a successful test of migrating a managed thing to this version? Scenario:
- Run the old version of the binding before this PR.
- Create a Thing using Main UI.
- Switch to the new version of the binding.
- The new channel should now appear in the already existing Thing and there should be no warnings logged (from the migration).

This would be very helpful to avoid any regressions.

Reply:
Those are the steps I did:
1. Removed my org.openhab.binding.mystrom.jar
2. Stopped OH
3. Commented out my mystrom.thing file
4. Started OH
5. Installed mystrom Binding in Main UI 
6. Added mystrom plug thing through Main UI
7. Uninstalled mystrom Binding
8. Installed org.openhab.binding.mystrom.jar in addons folder
9. The new channel did not appear 
10. Restarted OH
11. The new channel still did not appear
12. Only after uncommenting my thing in the mystrom.thing with another ID showed the new channel in the Main UI

It appears it´s not working after performing the steps you described. 
Besides that it´s working completely fine for me after step 12, probably because I only use the config files as configuration for my Things and Items.

I have also done this steps with changing the instructions in the instructions.xml from update to add, which I found on the documentation, but it seems this didnt help. (See my latest commit)

I also noticed that I cant add another mystrom Thing through the Main UI, since no mystrom Binding is showing in the list where you can select the Binding you want to create the Thing for.

Reply:
It wouldn't work with `<update-channel>`, but rather `<add-channel>` so that was a good finding. If you repeated all the steps after this change and it still didn't work, I guess there is some other problem. Was anything logged? You do not have to restart openHAB to test this. Since you are normally using things files, this is an advantage here, so I would rewrite the test scenario like this (only relative paths for /usr/share/openhab and /etc/openhab):
- Rename `addons/org.openhab.binding.mystrom.jar` to `org.openhab.binding.mystrom.jar`.
- Rename `things/mystrom.things` to `things/mystrom.things_`.
- Install binding through Main UI.
- Create a Thing using Main UI.
- Uninstall binding through Main UI.
- Rename `org.openhab.binding.mystrom.jar` to `addons/org.openhab.binding.mystrom.jar`.
- The new channel should now appear in the already existing Thing and there should be no warnings logged (from the migration).

And cleanup after tests:
- Delete Thing through Main UI.
- Rename `things/mystrom.things_` to `things/mystrom.things`.

Reply:
Thanks for the instructions. I did them but it seems it didnt work. I also changed the log level to `info` through the karaf console beforeheand. Here is the log with annotations what I did:

```
// remvoed addons/org.openhab.binding.mystrom.jar to deactivate it

2023-07-29 15:28:01.446 [INFO ] [org.apache.felix.fileinstall        ] - Uninstalling bundle 301 (org.openhab.binding.mystrom)

// installed Binding through MainUI

2023-07-29 15:29:28.059 [INFO ] [internal.service.FeaturesServiceImpl] - Adding features: openhab-binding-mystrom/[3.2.0,3.2.0]
2023-07-29 15:29:41.346 [INFO ] [internal.service.FeaturesServiceImpl] - Changes to perform:
2023-07-29 15:29:41.350 [INFO ] [internal.service.FeaturesServiceImpl] -   Region: root
2023-07-29 15:29:41.353 [INFO ] [internal.service.FeaturesServiceImpl] -     Bundles to install:
2023-07-29 15:29:41.358 [INFO ] [internal.service.FeaturesServiceImpl] -       mvn:org.openhab.addons.bundles/org.openhab.binding.mystrom/3.2.0
2023-07-29 15:29:41.368 [INFO ] [internal.service.FeaturesServiceImpl] - Installing bundles:
2023-07-29 15:29:41.371 [INFO ] [internal.service.FeaturesServiceImpl] -   mvn:org.openhab.addons.bundles/org.openhab.binding.mystrom/3.2.0
2023-07-29 15:29:41.497 [INFO ] [internal.service.FeaturesServiceImpl] - Starting bundles:
2023-07-29 15:29:41.504 [INFO ] [internal.service.FeaturesServiceImpl] -   org.openhab.binding.mystrom/3.2.0
2023-07-29 15:29:41.539 [INFO ] [internal.service.FeaturesServiceImpl] - Done.

// added mystrom Thing through MainUI

2023-07-29 15:31:02.353 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNINITIALIZED to INITIALIZING
2023-07-29 15:31:02.371 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from INITIALIZING to UNKNOWN
2023-07-29 15:31:02.480 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNKNOWN to ONLINE

// uninstalled Binding through MainUI

2023-07-29 15:31:51.656 [INFO ] [internal.service.FeaturesServiceImpl] - Removing features: openhab-binding-mystrom/[3.2.0,3.2.0]
2023-07-29 15:31:56.525 [INFO ] [internal.service.FeaturesServiceImpl] - Changes to perform:
2023-07-29 15:31:56.530 [INFO ] [internal.service.FeaturesServiceImpl] -   Region: root
2023-07-29 15:31:56.534 [INFO ] [internal.service.FeaturesServiceImpl] -     Bundles to uninstall:
2023-07-29 15:31:56.538 [INFO ] [internal.service.FeaturesServiceImpl] -       org.openhab.binding.mystrom/3.2.0
2023-07-29 15:31:56.541 [INFO ] [internal.service.FeaturesServiceImpl] - Stopping bundles:
2023-07-29 15:31:56.546 [INFO ] [internal.service.FeaturesServiceImpl] -   org.openhab.binding.mystrom/3.2.0
2023-07-29 15:31:56.579 [INFO ] [internal.service.FeaturesServiceImpl] - Uninstalling bundles:
2023-07-29 15:31:56.584 [INFO ] [internal.service.FeaturesServiceImpl] -   org.openhab.binding.mystrom/3.2.0
2023-07-29 15:31:56.613 [INFO ] [internal.service.FeaturesServiceImpl] - Refreshing bundles:
2023-07-29 15:31:56.616 [INFO ] [internal.service.FeaturesServiceImpl] -     org.openhab.binding.mystrom/3.2.0 (Bundle will be uninstalled)
2023-07-29 15:31:56.630 [INFO ] [internal.service.FeaturesServiceImpl] - Done.

2023-07-29 15:31:56.556 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from ONLINE to UNINITIALIZED
2023-07-29 15:31:56.572 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)

// added addons/org.openhab.binding.mystrom.jar 

2023-07-29 15:34:00.561 [INFO ] [org.apache.felix.fileinstall        ] - Installing bundle org.openhab.binding.mystrom / 4.0.0.202307290121
2023-07-29 15:34:00.675 [INFO ] [org.apache.felix.fileinstall        ] - Started bundle: file:/usr/share/openhab/addons/org.openhab.binding.mystrom-4.0.0-SNAPSHOT.jar
2023-07-29 15:34:00.714 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING
2023-07-29 15:34:00.728 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from INITIALIZING to UNKNOWN
2023-07-29 15:34:00.851 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNKNOWN to ONLINE

// the new channel did not appear in the UI
```
It seems the logs arent showing why the adding of the new channel wouldnt work. 
Do you maybe have another hint, what I could do?

Reply:
Thanks for the instructions. I did them but it seems it didnt work. I also changed the log level to `info` through the karaf console beforeheand. Here is the log with annotations what I did:

```
// remvoed addons/org.openhab.binding.mystrom.jar to deactivate it

2023-07-29 15:28:01.446 [INFO ] [org.apache.felix.fileinstall        ] - Uninstalling bundle 301 (org.openhab.binding.mystrom)

// installed Binding through MainUI

2023-07-29 15:29:28.059 [INFO ] [internal.service.FeaturesServiceImpl] - Adding features: openhab-binding-mystrom/[3.2.0,3.2.0]
2023-07-29 15:29:41.346 [INFO ] [internal.service.FeaturesServiceImpl] - Changes to perform:
2023-07-29 15:29:41.350 [INFO ] [internal.service.FeaturesServiceImpl] -   Region: root
2023-07-29 15:29:41.353 [INFO ] [internal.service.FeaturesServiceImpl] -     Bundles to install:
2023-07-29 15:29:41.358 [INFO ] [internal.service.FeaturesServiceImpl] -       mvn:org.openhab.addons.bundles/org.openhab.binding.mystrom/3.2.0
2023-07-29 15:29:41.368 [INFO ] [internal.service.FeaturesServiceImpl] - Installing bundles:
2023-07-29 15:29:41.371 [INFO ] [internal.service.FeaturesServiceImpl] -   mvn:org.openhab.addons.bundles/org.openhab.binding.mystrom/3.2.0
2023-07-29 15:29:41.497 [INFO ] [internal.service.FeaturesServiceImpl] - Starting bundles:
2023-07-29 15:29:41.504 [INFO ] [internal.service.FeaturesServiceImpl] -   org.openhab.binding.mystrom/3.2.0
2023-07-29 15:29:41.539 [INFO ] [internal.service.FeaturesServiceImpl] - Done.

// added mystrom Thing through MainUI

2023-07-29 15:31:02.353 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNINITIALIZED to INITIALIZING
2023-07-29 15:31:02.371 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from INITIALIZING to UNKNOWN
2023-07-29 15:31:02.480 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNKNOWN to ONLINE

// uninstalled Binding through MainUI

2023-07-29 15:31:51.656 [INFO ] [internal.service.FeaturesServiceImpl] - Removing features: openhab-binding-mystrom/[3.2.0,3.2.0]
2023-07-29 15:31:56.525 [INFO ] [internal.service.FeaturesServiceImpl] - Changes to perform:
2023-07-29 15:31:56.530 [INFO ] [internal.service.FeaturesServiceImpl] -   Region: root
2023-07-29 15:31:56.534 [INFO ] [internal.service.FeaturesServiceImpl] -     Bundles to uninstall:
2023-07-29 15:31:56.538 [INFO ] [internal.service.FeaturesServiceImpl] -       org.openhab.binding.mystrom/3.2.0
2023-07-29 15:31:56.541 [INFO ] [internal.service.FeaturesServiceImpl] - Stopping bundles:
2023-07-29 15:31:56.546 [INFO ] [internal.service.FeaturesServiceImpl] -   org.openhab.binding.mystrom/3.2.0
2023-07-29 15:31:56.579 [INFO ] [internal.service.FeaturesServiceImpl] - Uninstalling bundles:
2023-07-29 15:31:56.584 [INFO ] [internal.service.FeaturesServiceImpl] -   org.openhab.binding.mystrom/3.2.0
2023-07-29 15:31:56.613 [INFO ] [internal.service.FeaturesServiceImpl] - Refreshing bundles:
2023-07-29 15:31:56.616 [INFO ] [internal.service.FeaturesServiceImpl] -     org.openhab.binding.mystrom/3.2.0 (Bundle will be uninstalled)
2023-07-29 15:31:56.630 [INFO ] [internal.service.FeaturesServiceImpl] - Done.

2023-07-29 15:31:56.556 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from ONLINE to UNINITIALIZED
2023-07-29 15:31:56.572 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)

// added addons/org.openhab.binding.mystrom.jar 

2023-07-29 15:34:00.561 [INFO ] [org.apache.felix.fileinstall        ] - Installing bundle org.openhab.binding.mystrom / 4.0.0.202307290121
2023-07-29 15:34:00.675 [INFO ] [org.apache.felix.fileinstall        ] - Started bundle: file:/usr/share/openhab/addons/org.openhab.binding.mystrom-4.0.0-SNAPSHOT.jar
2023-07-29 15:34:00.714 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING
2023-07-29 15:34:00.728 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from INITIALIZING to UNKNOWN
2023-07-29 15:34:00.851 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'mystrom:mystromplug:7a4aadb190' changed from UNKNOWN to ONLINE

// the new channel did not appear in the UI
```
It seems the logs arent showing why the adding of the new channel wouldnt work. 
Do you maybe have another hint, what I could do?

Edit: I have used the .jar with update-channel and with add-channel in the instructions.xml - both have the same result

Reply:
Also want to add that I am using the following command for the built, which I found in the OH forum after encountering errors during the built:
clean install -Dohc.version=3.2.0 -Dmaven.compiler.source=11 -Dmaven.compiler.target=11 -Dkaraf.version=4.3.7 -pl :org.openhab.binding.mystrom

Reply:
So you are not testing this on a 4.0 system?

Reply:
I just saw that there is a new major version, I seem to have missed it completely. I am currently still on 3.2 as I had problems with another binding when I updated to 3.4. I have since left the system as is so that everything works properly. Is this going to be the problem? If so, I am sorry. I will tackle the update soon then.

Reply:
Yes, the upgrade mechanism was introduced in 4.0, so that explains why your test results were negative.

Please ping me in this PR when you have been able to test with 4.0, and I will get back to this.

If you are not ready to upgrade your production system just yet, another possibility is to install openHAB 4.0 on your development machine or any machine of your choice. In Windows it's a matter of minutes to download the distribution and run it.

Reply:
Hey again, I have done the steps on a new Windows runtime and it worked with the add-channel. So the channel is being added after installing the new version of the Binding.

> Can you confirm a successful test of migrating a managed thing to this version?
Now I can confirm this on OH 4.0.1

Reply:
Awesome, let's get it merged then - thanks!